### PR TITLE
Add API resource collection API

### PR DESCRIPTION
**Pull request by PasinduYeshan**
Please refer the [API Resource Collections Swagger](https://app.swaggerhub.com/apis/PasinduYeshan/wso-2_identity_server_api_resource_management_rest_api/v1#/API%20Resource%20Collections) section.

## Notes
- Merge https://github.com/wso2/carbon-identity-framework/pull/5137 before this PR.

## Related Issues 
- https://github.com/wso2/product-is/issues/17718

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/gen/java/org/wso2/carbon/identity/api/server/api/resource/v1/factories/ApiResourceCollectionsApiServiceFactory.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.api.server.api.resource.v1.factories;
+
+import org.wso2.carbon.identity.api.server.api.resource.v1.ApiResourceCollectionsApiService;
+import org.wso2.carbon.identity.api.server.api.resource.v1.impl.ApiResourceCollectionsApiServiceImpl;
+
+public class ApiResourceCollectionsApiServiceFactory {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/gen/java/org/wso2/carbon/identity/api/server/api/resource/v1/factories/ApiResourceCollectionsApiServiceFactory.java:24]: missing class comment

Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/gen/java/org/wso2/carbon/identity/api/server/api/resource/v1/factories/ApiResourceCollectionsApiServiceFactory.java:24]: extend AbstractFactoryBean.

check other factories

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/resources/APIResources.yaml [hunk 2]
```diff
@@ -629,7 +721,7 @@ components:
           example: before
         href:
           type: string
-          example: /o/orgName/api-resources?after=NDoy
+          example: /t/carbon.super/api/server/v1/api-resources?after=NDoy
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/resources/APIResources.yaml:724]:
```suggestion
          example: /api/server/v1/api-resources?after=NDoy
```
/t/carbon.super/ will not return now. /t/{tennat-domain} will be returned in other tenants. For this example we don't need to add it

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/util/APIResourceMgtEndpointUtil.java [hunk 3]
```diff
@@ -166,6 +168,21 @@ public static APIError handleAPIResourceMgtException(APIResourceMgtException e)
...
+    public static APIError handleAPIResourceCollectionMgtException(APIResourceCollectionMgtException e) {
+
+        Response.Status status = Response.Status.INTERNAL_SERVER_ERROR;
+        if (e instanceof APIResourceCollectionMgtClientException) {
+            LOG.debug(e.getMessage(), e);
+            status = Response.Status.BAD_REQUEST;
+        } else {
+            LOG.error(e.getMessage(), e);
+        }
+        String errorCode = e.getErrorCode(); //TODO: Update this
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/util/APIResourceMgtEndpointUtil.java:180]: Is this TODO needed

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    /**
+     * Build API Resource read write map.
+     *
+     * @param apiResourceCollection API Resource Collection.
+     * @return API Resource read write map.
+     */
+    private APIResourceMap buildAPIResourceMap(
+            APIResourceCollection apiResourceCollection) {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java:226]: can't we take these into 1 line

Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java]: Changed

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/util/APIResourceMgtEndpointUtil.java (hunk 3 above)
Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/util/APIResourceMgtEndpointUtil.java]: No this error codes were defined in framework. TODO is removed.

File: components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                validateRequiredAttributes(requestedAttributeList);
+            }
+
+            APIResourceCollectionSearchResult apiResourceCollectionSearchResult =
+                    APIResourceManagementServiceHolder.getApiResourceCollectionManager()
+                            .getAPIResourceCollections(filter, requestedAttributeList,
+                                    CarbonContext.getThreadLocalCarbonContext().getTenantDomain());
+            List<APIResourceCollection> apiResourceCollections =
+                    apiResourceCollectionSearchResult.getAPIResourceCollections();
+            if (apiResourceCollections == null || apiResourceCollections.isEmpty()) {
```
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java:80]: use collection utils for these cases

Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.api.server.api.resource/org.wso2.carbon.identity.api.server.api.resource.v1/src/main/java/org/wso2/carbon/identity/api/server/api/resource/v1/core/ServerAPIResourceCollectionManagementService.java]: Fixed

**Review by sadilchamishka — APPROVED** [review]

**Comment by sadilchamishka** [comment]
Let's fix the review comments added for this PR in a different PR Based on the urgency.

**Pull request merged**